Bugzilla/Template/Namespace.pm

```diff
@@ -0,0 +1,106 @@
+
+use base qw(Template::Base);
+use List::MoreUtils qw(any);
+
+my @BUGZILLA_HASHES    = qw( request_cache process_cache params localconfig );
+my @BUGZILLA_OBJECTS   = qw( cgi user );
+
+use Bugzilla::Template::Directive qw(tt_parse_ident tt_quote tt_dot_op);
+
+# These are also defined in a template, but for BMO we just want to fold them.
+my %TERMS = (
+    "bug"               => "bug",
+    "Bug"               => "Bug",
+    "abug"              => "a bug",
+    "Abug"              => "A bug",
+    "aBug"              => "a Bug",
+    "ABug"              => "A Bug",
+    "bugs"              => "bugs",
+    "Bugs"              => "Bugs",
+    "zeroSearchResults" => "Zarro Boogs found",
+    "Bugzilla"          => "Bugzilla",
+    "BugzillaTitle" => 'Bugzilla@Development',
+);
+
+sub ident {
+    my ($self, $ident) = @_;
+    my ($ns, $ns_args) = tt_parse_ident($ident);
+
+    my $expr;
+    if ($ns eq 'terms') {
+        my ($key, $key_args) = eval { tt_parse_ident($ident) };
+        if ($@) {
+            return Bugzilla::Template::Directive->ident([ tt_quote('terms'), 0, @$ident ]);
+        }
+        die "Unknown terms variable: terms.$key" unless exists $TERMS{$key};
+        return tt_quote($TERMS{$key});
+    }
+    elsif ($ns eq 'Hook') {
+        my ($method, $args) = tt_parse_ident($ident);
+        die "Bad hook method: $method" unless $method eq 'process';
+        $expr = "Bugzilla::Hook::template_process(\$context, $args)";
+    }
+    elsif ($ns eq 'constants') {
+        my ($const, $const_args) = tt_parse_ident($ident);
+        die "constants has args!" if $const_args;
+        $expr = "Bugzilla::Constants::$const";
+        if (my $f = Bugzilla::Constants->can($const)) {
+            my $val;
+            my $count = ($val) = $f->();
+            if ($count > 1) {
+                $expr = "[ $expr ]";
+            } 
+            elsif (ref $val && ref $val eq 'HASH') {
+                if (@$ident) {
+                    my ($key) = tt_parse_ident($ident);
+                    $expr .= "->{$key}";
+                }
+            }
+        }
+        # perl will catch invalid constants for us.
+    }
+    elsif ($ns eq 'Bugzilla') {
+        my ($method, $args) = tt_parse_ident($ident);
+        if (!Bugzilla->can($method)) {
+            die "No such method: Bugzilla->$method";
+        }
+        $expr = "Bugzilla->$method";
+
+        if ($args) {
+            $expr .= "($args)";
+        }
+        elsif (any { $method eq $_ } @BUGZILLA_HASHES) {
+            $expr .= '->{' . $ident->[0] . '}';
+            splice @$ident, 0, 2;
+        }
+        elsif (any { $method eq $_ } @BUGZILLA_OBJECTS) {
+            if (@$ident) {
+                my ($obj_method, $obj_args) = tt_parse_ident($ident);
+                $expr .= "->$obj_method";
+                if ($obj_args) {
+                    $expr .= "($obj_args)";
+                }
+            }
+        }
+    }
+
+    return tt_dot_op($expr, $ident);
+}
+
+
```

Comment:
Is this block of whitespace intentional? (As in, "missing code possibility" intentional, though "meets this community's whitespace specifications" is also a fair question.)